distributed/diagnostics/progress.py

```diff
@@ -291,3 +295,111 @@ def transition(self, key, start, finish, *args, **kwargs):
     def restart(self, scheduler):
         self.all.clear()
         self.state.clear()
+
+
+class GroupTiming(SchedulerPlugin):
+    """Keep track of high-level timing information for task group progress"""
+
+    name = "group-timing"
+
+    def __init__(self, scheduler):
+        scheduler.add_plugin(self)
+        self.scheduler = scheduler
+
+        # Timestamps for tracking compute durations by task group.
+        self.time: list[float] = [time.time()]
+        # The number of threads for the scheduler, used when estimating how occupied
+        # the workers are.
+        self.nthreads: list[int] = [self.scheduler.total_nthreads]
+        # The amount of compute since the last timestamp
+        self.compute: dict[str, list[float]] = {}
+
+        # Tracking the previous compute time values for getting deltas.
+        self._prev_durations: dict[str, dict[str, float]] = {}
+
+        for name, group in self.scheduler.task_groups.items():
+            self.compute[name] = [0.0]
+            self._prev_durations[name] = group.all_durations.copy()
+
+        # Start a callback sampling how the computations are going.
+        # TODO: this samples every 1s. Should it be configurable?
+        group_timing_pc = PeriodicCallback(self.track_groups, 1.0 * 1000.0)
```

Comment:
I pushed fairly hard on doing this on a per-task-group basis, and found it to be ultimately very difficult to do, for the reason that the samples are more interdependent than it might first seem.

The basic problem with it is that the \delta compute for each group might be (and very often is) larger than the time delta since the last checkpoint. This happens if some tasks take longer than each checkpoint, and can occur with either sampling strategy. In that case, we want to revise backwards the amount of compute and spread it out to previous checkpoints. Otherwise you wind up with workers that are several hundred percent occupied, even though in reality they have just occupied for a longer time.

Even worse, the different task groups are not independent of each other, and so you really want the sum of the groups over any given time period to be less than 100% of the available time. If they are on different time series, there needs to be a substantial amount of interpolation to get them aligned. This interpolation was getting both complex and computationally expensive. It starts to look a bit like solving a discretization-then-iterative-diffusion equation every render, which smelled very bad.

Finally, I'm going to argue that the one second sampling is not so much for the benefit of the task group progress tracking, but for the ultimate visualization. Say the dynamic range of computation durations is hugely different from the *plot discretization*: you will see a similar interpolation error with per-task-group tracking that we get with just doing time sampling. That is to say, I don't think the ultimate visualization will look very different. Now, a limitation here is if you wanted to start doing other things with the sampling data, e.g., show a high-resolution view on a single group. But for this application, I'm not sure we buy much with per-task-group tracking plus a lot of interpolation.